### ISSUE TITLE
C is incompatible with TypeVar(bound='C')

### ISSUE THREAD

**Issue by sk-** [python/mypy#5825]
The following code does not typecheck:

```python
from typing import List, TypeVar

_C = TypeVar('_C', bound='C')

class C:
  pass
def foo(a: List[_C]):
  a.append(C())  # incompatible types
def bar(a: List[C]):
  a.append(C())

def make_c_generic() -> _C:
  return C()  # incompatible types
def make_c() -> C:
  return C()
```

The error is:
```
list_type_var.py:8: error: Argument 1 to "append" of "list" has incompatible type "C"; expected "_C"
list_type_var.py:13: error: Incompatible return value type (got "C", expected "_C")
```

Note that the code is a simplified version of a code using generics which has the same problem.

Mypy version: 0.641
Mypy ini:
```
[mypy]
# This flag should be removed here and enabled per module when we have a
# considerable number of stubs for external libraries.
ignore_missing_imports = True
strict_optional = True
warn_unused_ignores = True
warn_redundant_casts = True
warn_unused_configs = True
check_untyped_defs = True
```

**Comment by Michael0x2a** [comment]
I think this is a legitimate error in your code -- see [invariance vs covariance](https://mypy.readthedocs.io/en/latest/common_issues.html#invariance-vs-covariance).

For example, suppose your `foo` function type-checked. That would end up causing the following code to break:

```python
class C:
    pass

class Child(C):
    pass

def foo(a: List[_C]):
  a.append(C())

children: List[Child] = [Child(), Child(), Child()]
foo(children)  # Type-checks, but now 'children' unexpectedly contains a C
```

If you don't actually need to mutate `a`, you could perhaps replace `List[_C]` with an immutable (and therefore covariant) collection like `Sequence[_C]`.

Similar-ish thing with `make_c_generic` -- if mypy decides to bind the typevar to some subclass of C in a particular context, returning a instance of C would be incorrect. You could maybe work around this by modifying the signature so the caller needs to pass in the subclass that needs to be instantiated -- e.g. pass in some value of type `Type[_C]` and instantiate that instead.

**Comment by ilevkivskyi** [comment]
Yes, this is not a bug, but actual type error.
